telegram/bot.py

```diff
@@ -82,13 +83,49 @@ class Bot(TelegramObject):
             :obj:`telegram.utils.request.Request`.
         private_key (:obj:`bytes`, optional): Private key for decryption of telegram passport data.
         private_key_password (:obj:`bytes`, optional): Password for above private key.
+        default_parse_mode (:obj:`str`, optional): Default parse mode used if not set explicitly in
+            method call. See the constants in :class:`telegram.ParseMode` for the available modes.
 
     """
 
+    def __new__(cls, *args, **kwargs):
+        # Handle default_... kwargs for bot methods
+        # Transform default_x=y kwargs into Defaults.x=y
+        defaults = Defaults()
+        for kwarg in list(kwargs.keys()):
+            if kwarg.startswith('default_'):
+                setattr(defaults, kwarg[8:], kwargs[kwarg])
+                del kwargs[kwarg]
+
+        # Make an instance of the class
+        instance = super(Bot, cls).__new__(cls)
+
+        # For each method ...
+        for method_name, method in inspect.getmembers(instance, predicate=inspect.ismethod):
+            # ... get kwargs
+            argspec = inspect.getargspec(method)
+            kwarg_names = argspec.args[-len(argspec.defaults or []):]
+            # ... check if Defaults has a attribute that matches the kwarg name
+            needs_default = [
+                kwarg_name for kwarg_name in kwarg_names if kwarg_name in defaults.__dict__.keys()
+            ]
+            # ... make a dict of kwarg name and the default value
+            default_kwargs = {
+                kwarg_name: getattr(defaults, kwarg_name) for kwarg_name in needs_default
+            }
+            # ... apply the defaults using a partial
+            if default_kwargs:
+                setattr(instance, method_name, functools.partial(method, **default_kwargs))
+
+        return instance
+
     def __init__(self, token, base_url=None, base_file_url=None, request=None, private_key=None,
-                 private_key_password=None):
+                 private_key_password=None, default_parse_mode=None):
```

Comment:
This probably won't work cause `del kwargs[kwarg]` (line 98) deletes this kwarg if i'm not mistaken.

A better way instead of `self.defaults = Defaults(parse_mode=default_parse_mode)` below, might be to simply do a `instance.defaults = defaults` in `__new__` above.

Reply:
Actually it does. Running

```
from telegram import Bot
bot = Bot('TOKEN', default_parse_mode='Markdown')
print(bot.defaults.parse_mode)
```

prints `Markdown` as expected. If I understand correctly, `__new__` can not directly alter the arguments passed to `__init__`. Instead for this to work one would have to overwrite the `__call__` method or something (see [here](https://bytes.com/topic/python/answers/751865-modify-arguments-between-__new__-__init__)).

Why did you add the `del kwargs[kwarg]` in the first place, if I may ask? The described behaviour in mind I think we can savely delete it (all tests pass without it, too).

Reply:
Originally i had `instance = super(Bot, cls).__new__(cls, *args, **kwargs)` note the args and kwargs. so i had to strip out the kwargs that `object.__new__` didn't want. I would prefer to have it like that, as it should provide better integration with custom Bot classes as far as I understand (instead of just ignoring the args and kwargs we actually pass them up according to the MRO)

Reply:
Okay, I see your point about passing the arguments. A solution for the problem that `object` takes no parameters (mentioned in the dev chat) could be to overwrite `TelegramObjects` (which inherits from `object`) `__new__` method to

```
def __new__(cls, *args, **kwargs):
    return super(TelegramObject, cls).__new__(cls)
```

i.e. stop the passing of args here. This passes all our tests but I'm not enough of a pythonian to tell if this might cause trouble later …

Still, I don't really see why the `default_` args should be deleted (and none of the other) before passing them along to `TelegramObject`. What am I missing?

Reply:
Bot inherits from `object` and should continue to do so. I was under the impression that it would be possible to override the MRO so that super would go up to a user defined class that does accept `__new__` but now that i think of it, that's not possible.
So as long as Bot inherits from object there's no need to pass the arguments.